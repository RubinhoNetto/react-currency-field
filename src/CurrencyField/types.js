// @flow
import * as React from 'react';

export type Mask = {
  numberToString: (num: string | number) => string,
  stringToNumber: (str: string) => number,
};

export type Values = {
  floatValue: number,
  formatedValue: string,
};

export type Locales = 'pt-BR' | 'en-US';

export type LocaleConfig = {
  charThousands: string,
  charDecimal: string,
  symbol: string,
  decimalScale: number,
};

export type Props = {
  value: string | number,
  name?: string,
  onChange?: (e: SyntheticEvent<HTMLInputElement>, values: Values) => void,
  onFocus?: (e: SyntheticEvent<HTMLInputElement>) => void,
  onBlur?: (e: SyntheticEvent<HTMLInputElement>) => void,
  className?: string,
  placeholder?: string,
  disabled?: boolean,
  required?: boolean,
  style?: { [string]: string | number },
  locale: Locales | LocaleConfig,
};
