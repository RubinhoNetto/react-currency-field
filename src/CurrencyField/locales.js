// @flow
export const defaultLocale = 'en-US';

export const locales = {
  'pt-BR': {
    charThousands: '.',
    charDecimal: ',',
    symbol: 'R$',
    decimalScale: 2,
  },
  'en-US': {
    charThousands: ',',
    charDecimal: '.',
    symbol: '$',
    decimalScale: 2,
  },
};
