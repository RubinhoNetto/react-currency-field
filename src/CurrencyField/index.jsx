// @flow
import * as React from 'react';
import moneyMask from './mask';
import type { Props, Values } from './types';
import { defaultLocale } from './locales';

let mask;

class CurrencyField extends React.Component<Props> {
  static defaultProps = {
    locale: defaultLocale,
    value: 0.0,
  };

  constructor(props: Props) {
    super(props);
    mask = moneyMask(this.props.locale);
  }

  formatValue = (value: string): Values => ({
    floatValue: mask.stringToNumber(value),
    formatedValue: mask.numberToString(value),
  });

  handleChange = (e: SyntheticEvent<HTMLInputElement>): void => {
    this.props.onChange &&
      this.props.onChange(e, this.formatValue(e.currentTarget.value));
  };

  getInputProps = () => {
    const { value, ...input } = this.props;

    delete input.locale;

    return {
      ...input,
      value: mask.numberToString(value),
      onChange: this.handleChange,
      type: 'text',
    };
  };

  render() {
    return <input {...this.getInputProps()} />;
  }
}

export default CurrencyField;
