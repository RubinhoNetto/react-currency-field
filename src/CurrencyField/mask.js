import { defaultLocale, locales } from './locales';

const mask = locale => {
  let options = locales[defaultLocale];

  if (typeof locale === 'string' && locales.hasOwnProperty(locale)) {
    options = locales[locale];
  } else if (typeof locale === 'object') {
    options = locale;
  }

  const initValue = (0.0).toFixed(options.decimalScale);

  const isNumber = num => Number(num) === num;

  const isValidStr = str => str.length >= options.decimalScale;

  const joinString = str =>
    `${str.substr(0, str.length - options.decimalScale)}.${str.substr(
      options.decimalScale * -1
    )}`;

  const transformStr = str =>
    parseInt(str)
      ? parseInt(str) / Math.pow(10, options.decimalScale)
      : parseFloat(initValue);

  const clearString = str => {
    const onlyNumbers = str.replace(/[\D]+/g, '');

    if (!isValidStr(onlyNumbers) && transformStr(onlyNumbers)) {
      return clearString(transformStr(onlyNumbers).toString());
    }
    return !!onlyNumbers ? onlyNumbers : clearString(initValue);
  };

  return {
    numberToString: num => {
      const number = clearString(
        isNumber(num) ? num.toFixed(options.decimalScale) : num
      );

      const formatNumber = joinString(number).split('.');
      formatNumber[0] = formatNumber[0]
        .split(/(?=(?:...)*$)/)
        .join(options.charThousands);

      return `${options.symbol} ${formatNumber.join(options.charDecimal)}`;
    },

    stringToNumber: str => parseFloat(joinString(clearString(str))),
  };
};

export default mask;
