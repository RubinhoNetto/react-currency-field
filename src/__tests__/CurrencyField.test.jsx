import * as React from 'react';
import * as ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import CurrencyField from '../CurrencyField';

enzyme.configure({ adapter: new Adapter() });

const props = {
  value: 0.0,
};

const customLocale = {
  charThousands: '-',
  charDecimal: '_',
  symbol: '#',
  decimalScale: 2,
};

const simulate = (value, expected, locale = 'en-US') => {
  let formated = 0;

  mount(
    <CurrencyField
      locale={locale}
      value={value}
      onChange={(e, v) => {
        formated = v.formatedValue;
      }}
    />
  ).simulate('change');

  expect(formated).toEqual(expected);
};

describe('Currency Field Tests', () => {
  describe('Default Tests', () => {
    it('should render a default Currency Field', () => {
      ReactDOM.render(
        <CurrencyField {...props} />,
        document.createElement('div')
      );
    });

    it('should match the default snapshot', () => {
      const field = renderer.create(<CurrencyField {...props} />);
      expect(field.toJSON()).toMatchSnapshot();
    });

    it('should has a locale `en-US` by default', () => {
      const field = mount(<CurrencyField {...props} />);
      expect(field.prop('locale')).toEqual('en-US');
    });

    it('should set locale as `pt-BR`', () => {
      const field = mount(<CurrencyField {...props} locale="pt-BR" />);
      expect(field.prop('locale')).toEqual('pt-BR');
    });

    it('should only allow type numbers', () => {
      simulate('money', '$ 0.00');
    });

    it('should has all passed props', () => {
      const field = mount(
        <CurrencyField
          value={10.5}
          className="my-class"
          id="my-id"
          required
          disabled
        />
      );
      const p = field.find('input').props();

      expect(p.value).toEqual('$ 10.50');
      expect(p.className).toEqual('my-class');
      expect(p.id).toEqual('my-id');
      expect(p.required).toEqual(true);
      expect(p.disabled).toEqual(true);
    });
  });

  describe('Tests of the mask in `en-US`', () => {
    it('Type `22.3` and expect the value to be `$ 22.30`', () => {
      simulate(22.3, '$ 22.30');
    });

    it('Type `2456.1` and expect the value to be `$ 2,456.10`', () => {
      simulate(2456.1, '$ 2,456.10');
    });

    it('Type `12` and expect the value to be `$ 12.00`', () => {
      simulate(12, '$ 12.00');
    });

    it('Type `$ 123.33` and expect the value to be `$ 123.33`', () => {
      simulate('$ 123.33', '$ 123.33');
    });

    it('Type `R$ 123,33` and expect the value to be `$ 123.33`', () => {
      simulate('R$ 123,33', '$ 123.33');
    });
  });

  describe('Tests of the mask in `pt-BR`', () => {
    it('Type `22.3` and expect the value to be `R$ 22,30`', () => {
      simulate(22.3, 'R$ 22,30', 'pt-BR');
    });

    it('Type `2456.1` and expect the value to be `R$ 2.456,10`', () => {
      simulate(2456.1, 'R$ 2.456,10', 'pt-BR');
    });

    it('Type `12` and expect the value to be `R$ 12,00`', () => {
      simulate(12, 'R$ 12,00', 'pt-BR');
    });

    it('Type `$ 123.33` and expect the value to be `R$ 123,33`', () => {
      simulate('$ 123.33', 'R$ 123,33', 'pt-BR');
    });

    it('Type `$ 123.33` and expect the value to be `R$ 123,33`', () => {
      simulate('$ 123.33', 'R$ 123,33', 'pt-BR');
    });
  });

  describe('Tests of the mask with a custom locale config', () => {
    it('Type `22.3` and expect the value to be `# 22_30`', () => {
      simulate(22.3, '# 22_30', customLocale);
    });

    it('Type `2456.1` and expect the value to be `# 2-456_10`', () => {
      simulate(2456.1, '# 2-456_10', customLocale);
    });
  });

  describe('Tests of the mask with a custom locale config', () => {
    it('Type `22.3` and expect the value to be `# 22_30`', () => {
      simulate(22.3, '# 22_30', customLocale);
    });

    it('Type `2456.1` and expect the value to be `# 2-456_10`', () => {
      simulate(2456.1, '# 2-456_10', customLocale);
    });
  });

  describe('Tests of the mask passing a string', () => {
    it('Type `$ 12,455.30` and expect the value to be `$ 12,455.30`', () => {
      simulate('$ 12,455.30', '$ 12,455.30');
    });

    it('Type `12455.30` and expect the value to be `$ 12,455.30`', () => {
      simulate('12455.30', '$ 12,455.30');
    });

    it('Type `125.51` and expect the value to be `$ 125.51`', () => {
      simulate('125.51', '$ 125.51');
    });

    it('Type `40` and expect the value to be `$ 40.00`', () => {
      simulate('40', '$ 40.00');
    });

    it('Type `0.05` and expect the value to be `$ 00.05`', () => {
      simulate('0.05', '$ 0.05');
    });

    it('Type `8` and expect the value to be `$ 8.00`', () => {
      simulate('8', '$ 8.00');
    });
  });
});
