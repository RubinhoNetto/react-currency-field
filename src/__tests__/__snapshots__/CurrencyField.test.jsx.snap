// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency Field Tests Default Tests should match the default snapshot 1`] = `
<input
  onChange={[Function]}
  type="text"
  value="$ 0.00"
/>
`;
